### ISSUE TITLE
Stopping the MobFox Service

### ISSUE THREAD

**Issue by iWareDesigns** [mobfox/MobFox-Android-SDK#58]
Hi,

While testing we started experiencing a crash after the following circumstances.

1. We launched our App and displayed some MobFox adverts.
2. We then closed down our App.
3. Re-launching the App causes a crash due to the MobFox Service.

We use an NDK shared library and on the Activity onDestroy() method we kill the process so that the shared library is reloaded when the App is re-launched:

```
//	As Android does not destroy the app properly, we should kill it!
ActivityManager activityManager = (ActivityManager)getSystemService(ACTIVITY_SERVICE);
activityManager.killBackgroundProcesses(Prophet.appContext.getPackageName());
```

Is there any functionality in the MobFox SDK to stop the service defined in the Manifest?

`<service android:name="com.mobfox.sdk.services.MobFoxService" android:launchMode="singleTop" />`

This is the Logcat output with when the App stops responding / crashes on re-launch.

```
E/ActivityManager: ANR in com.iwaredesigns.prosnooker2012
                                                PID: 29824
                                                Reason: executing service com.iwaredesigns.prosnooker2012/com.mobfox.sdk.services.MobFoxService
                                                Load: 5.71 / 5.99 / 5.6
                                                CPU usage from 654492ms to 0ms ago:
                                                  15% 189/surfaceflinger: 8.1% user + 7.3% kernel / faults: 350461 minor 1 major
                                                  6.3% 609/system_server: 4.5% user + 1.7% kernel / faults: 91070 minor 66 major
                                                  3.1% 194/mediaserver: 1.6% user + 1.5% kernel / faults: 2208 minor 83 major
                                                  3% 699/sdcard: 0% user + 2.9% kernel / faults: 40 minor
                                                  2.4% 27932/com.android.packageinstaller: 1.8% user + 0.5% kernel / faults: 13824 minor 43 major
                                                  1.7% 21604/com.android.vending: 1.3% user + 0.3% kernel / faults: 10738 minor 96 major
                                                  1.2% 203/mpdecision: 0.2% user + 1% kernel / faults: 42 minor
                                                  1.2% 1126/com.android.launcher: 0.9% user + 0.3% kernel / faults: 45469 minor 5 major
                                                  1% 716/com.android.systemui: 0.6% user + 0.3% kernel / faults: 19612 minor 19 major
                                                  0.8% 88/msmfb_commit_th: 0% user + 0.8% kernel
                                                  0.8% 1437/com.google.android.gms: 0.6% user + 0.2% kernel / faults: 68279 minor 80 major
                                                  0.7% 14743/kworker/u:4: 0% user + 0.7% kernel
                                                  0.6% 23332/com.android.defcontainer: 0.2% user + 0.4% kernel / faults: 2559 minor 12 major
                                                  0.6% 26916/kworker/u:0: 0% user + 0.6% kernel
                                                  0.4% 140/mmcqd/0: 0% user + 0.4% kernel
                                                  0.3% 151/ueventd: 0% user + 0.2% kernel
                                                  0.3% 1321/com.google.android.gms.persistent: 0.2% user + 0% kernel / faults: 10305 minor 46 major
                                                  0.2% 1561/com.estrongs.android.pop: 0.1% user + 0% kernel / faults: 19828 minor 27 major
                                                  0.2% 26928/kworker/u:8: 0% user + 0.2% kernel
                                                  0.2% 65/kswapd0: 0% user + 0.2% kernel
                                                  0.2% 27208/kworker/0:2: 0% user + 0.2% kernel
                                                  0.1% 1025/com.google.android.inputmethod.latin: 0.1% user + 0% kernel / faults: 5347 minor 34 major
                                                  0.1% 2220/com.facebook.katana: 0.1% user + 0% kernel / faults: 10352 minor 241 major
                                                  0.1% 156/flush-179:0: 0% user + 0.1% kernel
                                                  0.1% 829/RX_Thread: 0% user + 0.1% kernel
                                                  0.1% 26471/kworker/u:1: 0% user + 0.1% kernel
                                                  0.1% 27216/kworker/u:9: 0% user + 0.1% kernel
                                                  0.1% 26709/kworker/0:0: 0% user + 0.1% kernel
                                                  0% 26926/kworker/u:6: 0% user + 0% kernel
                                                  0% 188/servicemanager: 0% user + 0% kernel / faults: 57 minor
                                                  0% 26924/kworker/u:2: 0% user + 0% kernel
                                                  0% 195/installd: 0% user + 0% kernel / faults: 63 minor
                                                  0% 14909/com.google.android.googlequicksearchbox:search: 0% user + 0% kernel / faults: 8759 minor 247 major
                                                  0% 174/logd: 0% user + 0% kernel / faults: 102 minor
                                                  0% 828/TX_Thread: 0% user + 0% kernel
                                                  0% 6/migration/0: 0% user + 0% kernel
                                                  0% 730/android.process.media: 0% user + 0% kernel / faults: 1282 minor 2 major
                                                  0% 827/MC_Thread: 0% user + 0% kernel
                                                  0% 163/jbd2/mmcblk0p30: 0% user + 0% kernel
                                                  0% 980/wpa_supplicant: 0% user + 0% kernel / faults: 133 minor
                                                  0% 191/netd: 0% user + 0% kernel / faults: 1443 minor
                                                  0% 1107/com.android.phone: 0% user + 0% kernel / faults: 581 minor 2 major
                                                  0% 205/zygote: 0% user + 0% kernel / faults: 14129 minor
                                                  0% 187/lmkd: 0% user + 0% kernel / faults: 24 minor
                                                  0% 202/thermald: 0% user + 0% kernel / faults: 40 minor
                                                  0% 1079/com.android.nfc: 0% user + 0% kernel / faults: 451 minor
                                                  0% 2/kthreadd: 0% user + 0% kernel
                                                  0% 201/bridgemgrd: 0% user + 0% kernel / faults: 454 minor
                                                  0% 186/healthd: 0% user + 0% kernel
                                                  0% 1839/.esfm: 0% user + 0% kernel / faults: 431 minor
                                                  0% 23479/com.concretesoftware.pbachallenge_androidmarket: 0% user + 0% kernel / faults: 486 minor
                                                  0% 175/vold: 0% user + 0% kernel / faults: 34 minor
                                                  0% 23148/android.process.acore: 0% user + 0% kernel / faults: 1116 minor 14 major
                                                  0% 3/ksoftirqd/0: 0% user + 0% kernel
                                                  0% 23389/com.android.musicfx: 0% user + 0% kernel / faults: 1100 minor
                                                  0% 1//init: 0% user + 0% kernel / faults: 132 minor
                                                  0% 24/sync_supers: 0% user + 0% kernel
                                                  0% 27/msm_slim_ctrl_r: 0% user + 0% kernel
                                                  0% 110/irq/330-wcd9xxx: 0% user + 0% kernel
                                                  0% 193/drmserver: 0% user + 0% kernel / faults: 223 minor 3 major
                                                  0% 23403/com.android.keychain: 0% user + 0% kernel / faults: 312 minor
                                                  0% 23519/com.android.documentsui: 
```

Regards,

iWare Designs.

**Comment by iWareDesigns** [comment]
Hi,

I have just been checking the MobFox documentation and have a query.

Your DemoApp manifest describes the following service.

`<service android:name="com.mobfox.sdk.services.MobFoxService" android:launchMode="singleTop" />`

https://github.com/mobfox/MobFox-Android-SDK/blob/master/demoapp/src/main/AndroidManifest.xml

Your Integration guide does not mention anything about adding the service to the manifest.

I have tested the banner and interstitial adverts without the service entry in the manifest and they work 
 okay.

Is the addition of the service in the manifest specific to the DemoApp?

Regards,

iWare Designs.

**Comment by brachil** [comment]
Not sure why you are getting this crash, I tried to reproduce but did not manage, all was fine.
If you remove your lines in onDestroy - does the crash still happen?
Our service also has:
onDestroy()
{
 ....
 ......
 Process.killProcess(Process.myPid()); 
}
So maybe the duplication causes the crash. (you can try putting your kill in a try/catch might solve)
 
Installation guide does mention to add this service to manifest.
But as for now it is not mandatory and you can just leave the service out the manifest and it will not be called.

**Comment by iWareDesigns** [comment]
Hi,

We think the problem is being caused when Android does not call the onDestroy() method.  Apparently the onDestroy() method is not guaranteed to be called when an App is closed.

The safest option for us is to release without the MobFoxService, what impact will this have on delivered adverts?  Will this reduce the potential advert pool / eCPM's etc...?

Regards,

iWare Designs.

**Comment by nabriski** [comment]
At this point it will not have any impact. You can safely go on and remove it.

Regards,
Itamar

**Comment by iWareDesigns** [comment]
Excellent, thank you.

**Comment by iWareDesigns** [comment]
Hi,

Some more information about the MobFoxService crash that I have experienced.

When the App is closed and the Android system does not call the onDestroy() method the MobFoxService is not destroyed.

After about 5 to 10 seconds the Android system kills the MobFoxService as it is ANR (App Not Responding).

If we try to re-launch the App before the Android system kills the MobFoxService, the App seems to re-connect to the old MobFoxService and hangs, and then eventually displays an ANR dialog.

The whole problem seems to be caused by Android not called the onDestory() method under certain circumstances.  A possible solution may be stopping the service via the Activity onPause() method and restarting it on the Activity onResume() method.

We have noticed that the onDestroy() method is not guaranteed to be called when closing the App after pressing the "Overview" button and swiping the App off the screen to close it.  However in this instance the onPause() and onResume() methods are called.

Regards,

iWare Designs.
